File: c2_datatype/src/main/java/xx_kotlin/generic/TestOutIn.java
package xx_kotlin.generic;

import java.util.ArrayList;
import java.util.List;

public class TestOutIn {

    ParentClass parentClass=new ParentClass();
    ChildClass mChildClass=new ChildClass();

    void test01(){
        //error
//        List<ParentClass> pClasses=new ArrayList<ChildClass>();
        List<? extends ParentClass> pClasses=new ArrayList<ChildClass>();

        //? extends ParentClass> 只能获取
        // 不能修改(null除外)
//        pClasses.add(parentClass);
//        pClasses.add(mChildClass);
        pClasses.add(null);

        //只能获取
        ParentClass parentClass=pClasses.get(0);


        //? super ChildClass 能修改
        //不能获取
        List<? super ChildClass> cClasses=new ArrayList<ParentClass>();
//        cClasses.add(parentClass);
        cClasses.add(mChildClass);

//        ChildClass childClass= cClasses.get(0);
    }
}

File: c2_datatype/src/main/java/xx_kotlin/generic/TestOutIn.kt
package xx_kotlin.generic

val parentClass = ParentClass()
val childClass = ChildClass()

fun test01() {

    //out :向外拿的，获取
    //out 只能获取，不能修改   ? extend ParentClass
    val list: MutableList<out ParentClass> = ArrayList<ChildClass>()

    //in :向内放的，修改
    //in 只能修改，不能获取   ? super ChildClass
    val inList: MutableList<in ChildClass> = ArrayList<ParentClass>()


}


